### ISSUE TITLE
Privileged Identity Management (PIM)

### ISSUE THREAD

**Issue by Lachlan-White** [hashicorp/terraform-provider-azuread#68]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description
Ability to Create PIM Policies and Configure PIM access via Terraform

### New or Affected Resource(s)

New data source would be required 

* azuread_privileged_identity_management

### Potential Terraform Configuration

```hcl
resource "azurerm_priviliged_identity_management" "PIM-Group-1" {
  scope = "Subscription_PIM_1"
  role_definition_name = "Contributor"
  aad_group_id = "${var.aad_group_id}"
}
```

### References
https://docs.microsoft.com/en-us/azure/active-directory/privileged-identity-management/pim-configure

**Comment by mharrison365** [comment]
What would also be super helpful here would be support for "Azure Resources" PIM.

https://docs.microsoft.com/en-us/azure/active-directory/privileged-identity-management/pim-resource-roles-assign-roles

**Comment by manicminer** [comment]
Thanks for requesting this feature! It looks like PIM management is only available via the [MS Graph API](https://docs.microsoft.com/en-us/graph/api/resources/privilegedidentitymanagement-root?view=graph-rest-beta) which we are currently not in a position to use as we're [waiting on SDK support](https://github.com/Azure/azure-sdk-for-go/issues/7352). I've marked this request as upstream-dependent and we'll revisit as soon as we are able.

**Comment by khole** [comment]
maybe add the assignment type (eligible/active)
and assignment duration (start/end) 
as parameters?

**Comment by LudoNieuwenhuizen** [comment]
Would be super handy for our production setup. Currently, this missing functionality means quite some manual work

**Comment by manicminer** [comment]
This feature is currently waiting on implementation of MS Graph. We are actively working on it and this issue will receive priority as soon as we are able. See #323 for more details. Thanks!

**Comment by frazhamid** [comment]
A few use cases I would like to see met:
- Adding Subscriptions
- Configuring Role Settings for both Azure AD and Azure Resources
- Configuring Alerts
- Configuring Access Reviews
- Managing Role Assignments (Add/Update/Remove)

Presently I am resorting to using the AzureAD Preview PowerShell cmdlets. These work well but it is not as elegant as tf.

**Comment by darren-johnson** [comment]
We create all our RBAC groups using terraform when a resource group is provisioned, but they need manually onboarding to PIM so this would be very useful.

Any update?

**Comment by schlbra** [comment]
@manicminer , it appears that MS Graph was implemented in azuread version 1.5.0 [(LINK HERE)](https://github.com/hashicorp/terraform-provider-azuread/issues/323#issuecomment-845163725).

Does this mean that this issue can receive priority now or maybe it's already in progress?

**Comment by manicminer** [comment]
In order to maintain compatibility, new features will only be merged in v2.0 (or later), which we're working on that right now.

Privileged Identity Management is a reasonably large feature that will likely not make it into 2.0, but will hopefully come soon thereafter as it's the highest voted feature on our backlog.

**Comment by JonZeolla** [comment]
I'm already subscribed to this issue but whenever it's ready I'm happy to volunteer testing time.

**Comment by bodemckenna** [comment]
Can you provide any additional information about when this is likely to be released.

**Comment by Threpio** [comment]
I am interested in this feature as-well - Happy to implement the resources and datasources myself if help is needed.

@manicminer - Let me know by ping, this is required by my workplace aswell.

**Comment by tepley** [comment]
I'm also interested this feature.

**Comment by manicminer** [comment]
Thanks for the interest, but please stick to +1 on the original message. Help is always very welcome, however this particular feature is nontrivial - for context, there are [3 distinct iterations of this service](https://docs.microsoft.com/en-us/azure/active-directory/privileged-identity-management/pim-apis), incompatible with each other and none of which have yet made it out of beta.

**Comment by TomHowarth** [comment]
Please stick me as a Plus 1 on implementing this in Azuread

**Comment by masoud-fereidonian** [comment]
+1

**Comment by pbeckendorf-gkgab** [comment]
Dear All, as Microsoft changed it's PIM API in October (https://docs.microsoft.com/en-us/graph/api/resources/privilegedidentitymanagement-root?view=graph-rest-beta), it might be a good thing to ask once again how the Status for the Azure PIM Terraform Provider is at the Moment? Anything planed or is there nothing on the horizon? Just wondering, as i think this would be a great thing for AAD PIM and Azure Resources PIM to be able to enroll it with Terraform.

**Comment by RenatoMartins-tomtom** [comment]
This would be really helpful, as we are planning to onboard management group structure into PIM, and we keep the RBAC assignments on the MG scope in TF state.

**Comment by petr-stupka** [comment]
Seems like there is already work in progress - https://github.com/hashicorp/terraform-provider-azuread/issues/547

**Comment by ZarakiKenpachi7** [comment]
Hello, checking up on this one as we need it via tf rather than regular clickops. Any updates, please?

**Comment by archmangler** [comment]
Hi, what's the status of this much desired and valuable feature ?

**Comment by speedyankur** [comment]
We have found a work around for PIM implementation in terraform 
https://goodworkaround.com/2021/10/14/assigning-pim-azure-rbac-permissions-using-terraform-and-arm-template/

However Im still waiting on updates on this one. Any timelines..

**Comment by mariussm** [comment]
@manicminer All API endpoints for Azure AD roles are now out beta (and have been for a while): https://learn.microsoft.com/en-us/graph/api/resources/privilegedidentitymanagementv3-overview?view=graph-rest-1.0

Any chance this will be prioritized in the backlog anytime soon? 

I wish I had the required expertise for contributing to the provider, but I lack any Golang knowledge. If there is anything I can provide on the Graph side, or through testing, I am happy to help.

**Comment by MohnJadden** [comment]
This should really be prioritized - the workaround @speedyankur mentioned is not working for me, and I've gotten no response from its original author on their Github. We can do everything we need except for PIM, which is still a manual step since TF doesn't support it.

**Comment by thedevopsjedi** [comment]
Hi folks, I got fed up of waiting for this to be resolved, so I wrote a script that does it using the Azure APIs. This could easily be triggered from terraform or via a pipeline which is how I do it.  If you are interested the script is available at 
https://github.com/thedevopsjedi/azure-pim

**Comment by MohnJadden** [comment]
> 

Thank you for this - do you happen to have a guide on how to trigger via TF?  I couldn't find much that allows us to directly trigger a resource provider that runs Powershell scripts with TF variables as PS parameters.

**Comment by thedevopsjedi** [comment]
A guide specifically no, as I did it in a pipeline but its simple to do, example code below and I'd just use terraform's data sources and outputs for the PowerShell parameters.

```hcl
resource "null_resource" "trigger_powershell_script" {
  provisioner "local-exec" {
    command     = <<EOT
POWERSHELL COMMANDS GO HERE
EOT
    interpreter = ["pwsh", "-Command"]
  }
  depends_on = [RESOURCES]
}
```

**Comment by pckls** [comment]
Hi All,

https://learn.microsoft.com/en-us/azure/active-directory/fundamentals/whats-new#june-2023

This includes the following:
* General Availability - Privileged Identity Management and Conditional Access integration
* General Availability - Privileged Identity Management for Groups

I'm not quite sure on the API's, but they seem to be here:
* https://learn.microsoft.com/en-us/graph/api/resources/privilegedidentitymanagementv3-overview
* https://learn.microsoft.com/en-us/graph/api/resources/privilegedidentitymanagement-for-groups-api-overview

**Comment by michvllni** [comment]
Hi All,

this was implemented in the new Release of the azurerm provider: [v3.64.0](https://github.com/hashicorp/terraform-provider-azurerm/releases/tag/v3.64.0)

**Comment by devblackops** [comment]
@michvllni It looks like the new `azurerm_pim_active_role_assignment`, and `azurerm_pim_eligible_role_assignment` resources in the `azurerm` provider only partially implement PIM. It would still be nice to be able to onboard groups into PIM and manage their respective settings (is MFA or justification required, the list of approvers, etc.). It would make sense for those resources to be in the `azuread` provider, not `azurerm`. I've implemented my own IaC solution for this, but would much prefer if this were supported in the `azuread` provider natively.

**Comment by michvllni** [comment]
I completely agree with you that it is not complete, just wanted to give you the heads up that at least some part of PIM is now implemented

**Comment by MohnJadden** [comment]
I haven't had luck getting the current implementation to work - when I try to assign PIM eligible roles via a mapped object variable, 10 of the 11 PIM roles look like they'll get assigned without incident, one throws a weird error.  [I have an issue open in the azurerm resource provider Github](https://github.com/hashicorp/terraform-provider-azurerm/issues/22513), hopefully it gets worked on.  

> @michvllni It looks like the new `azurerm_pim_active_role_assignment`, and `azurerm_pim_eligible_role_assignment` resources in the `azurerm` provider only partially implement PIM. It would still be nice to be able to onboard groups into PIM and manage their respective settings (is MFA or justification required, the list of approvers, etc.). It would make sense for those resources to be in the `azuread` provider, not `azurerm`. I've implemented my own IaC solution for this, but would much prefer if this were supported in the `azuread` provider natively.

2nding the request to manage settings - an issue I have with PIM is that every RG has its own separate setting for eligibility duration.  If I want to make a resource group allow permanently eligible PIM assignments, I have to make that change for each and every resource group within PIM.  I'm sure there's a way to specify this via a PS script or API calls but this is a perfect use case for Terraform, either via azurerm or azuread resource provider.  

I too would love to get the IaC solution you were able to put together.

**Comment by MohnJadden** [comment]
Is there some way to get beta access or to be able to use the module that oWretch has a branch for with @manicminer ?  I'd love to test and see if they help my situation.

**Comment by oWretch** [comment]
Hi @MohnJadden. Getting this going in Terraform is a two step process. The above PR is to translate the Graph APIs into a common format that I can then use in the provider. Once that is done I’ll open a separate PR and link it here which you will be able to compile and use before it gets approved. 

What are you needing to do? I can try and make sure that I cover your needs in the development. My aim is to cover all the Group PIM APIs for now, which should mean enabling a group for PIM access, configuring the policies, and finally adding users or groups to the group either as eligible or active.

**Comment by unique-dominik** [comment]
> Hi All,
> 
> It looks like a few PIM features entered GA in the past week 😃
> 
> https://learn.microsoft.com/en-us/azure/active-directory/fundamentals/whats-new#june-2023
> 
> This includes the following:
> 
> * General Availability - Privileged Identity Management and Conditional Access integration
> * General Availability - Privileged Identity Management for Groups
> 
> I'm not quite sure on the API's, but they seem to be here:
> 
> * https://learn.microsoft.com/en-us/graph/api/resources/privilegedidentitymanagementv3-overview
> * https://learn.microsoft.com/en-us/graph/api/resources/privilegedidentitymanagement-for-groups-api-overview

Do I see correct that even with the ⬆️ addition (azurerm provider >3.64) you can not manage the role settings yet?
Or do I miss something?

<img width="688" alt="image" src="https://github.com/hashicorp/terraform-provider-azuread/assets/103249640/432f896f-1f1d-4ab8-9076-6758dedb831e">

**Comment by manicminer** [comment]
Hi @MohnJadden, thanks for reporting these issues with the current PIM implementation in AzureRM. Please could you open a bug report on the AzureRM repository and we'll be happy to look at what we can improve. Thanks!

**Comment by MohnJadden** [comment]
I've been opening issues like gangbusters since the resource provider came out in 3.65, [including one that details what happens when we try to assign a PIM eligible role without a schedule block](https://github.com/hashicorp/terraform-provider-azurerm/issues/22766) as described above.  Another user has [an issue open for Azure AD PIM role support as well](https://github.com/hashicorp/terraform-provider-azurerm/issues/22569).

**Comment by manicminer** [comment]
Awesome, thanks for those, apologies I hadn't seen them yet.

**Comment by jjaychen1e** [comment]
Any updates or plans about role settings (role management)?

**Comment by MohnJadden** [comment]
I'm still seeing the same issue in #22766 and we're on 3.70.  Who do I talk to about contributing to the azurerm module in the form of human-testing PIM functionality?

**Comment by sokoloffmaks** [comment]
Hi, wondering if this issue covers assigning PIM eligibility  to azure resource groups /resources https://learn.microsoft.com/en-us/azure/active-directory/privileged-identity-management/pim-resource-roles-configure-role-settings ?

**Comment by manicminer** [comment]
Hi all, your feedback is very much appreciated, but where reasonably practical can we please hold these discussions on the relevant issue thread. I realise this functionality in general crosses both providers but the preceding comments seem to be about the AzureRM provider implementation? Thanks

**Comment by MohnJadden** [comment]
@manicminer I've been doing just that in https://github.com/hashicorp/terraform-provider-azurerm/issues/22766 for several weeks now, including an @ asking you if there was any timeframe on addressing the standing issue there.  

Would it be easier on the TF team if this request was closed and new/duplicate requests were spawned - one for PIM for groups, one for Azure AD PIM?

**Comment by magodo** [comment]
> Hi, wondering if this issue covers assigning PIM eligibility to azure resource groups /resources https://learn.microsoft.com/en-us/azure/active-directory/privileged-identity-management/pim-resource-roles-configure-role-settings ?

There is a PR for this feature in the azurerm provider: https://github.com/hashicorp/terraform-provider-azurerm/pull/23295

**Comment by MohnJadden** [comment]
For terminology parts, the scope is where in Azure you want the PIM role to apply.  Let's say you have 40 clients with 40 Azure resource groups.  One client wants to be able to create and delete VMs in their RG, named BigClient, via the VM Contributor role.  Their user accounts are already in the Azure AD group "eligible-group".  

In the Azure Portal -> PIM -> Azure resources page, you navigate to the BigClient RG and click on it -> click on Settings.  You navigate to and click on VM Contributor and configure its Activation setting duration to 8 hours.  Then you go back to the RG view -> Assignments -> Add Assignment.  In the Select Role dropdown, select VM Contributor.  Click on Select Members -> Eligible Group.  Select eligible-group from the AAD groups and users list -> click Select.  Click Next -> set how long the group has the option to PIM to the role by setting an eligibility expiration date/time (you can remove the requirement to set an expiration in the Settings section we described above).  Next, Finish, done - members of that group can now activate PIM roles to take VM Contributor actions in the BigClient RG.  

Full disclosure, I've had my errors with the TF module, so I can't confirm it's 100% functional.  To mirror your use case with static everything, it would look like the example below.  Bear in mind that in order to get the role_definition_id, you'd need to go into IAM -> Roles and copy the GUID of the VM Contributor role and also the ID of the group from Azure AD -> Groups -> search for the group name.

```
resource "time_static" "starttime"{}

resource "azurerm_pim_eligible_role_assignment" "vm-contrib-bigclient" {
  scope                = "/subscriptions/theSubscriptionIDOfTheAzureSubscriptionWhichContainsTheResourceGroup/resourceGroups/BigClient"
  role_definition_id = "/subscriptions/theSubscriptionIDOfTheAzureSubscriptionWhichContainsTheResourceGroup/providers/Microsoft.Authorization/roleDefinitions/TheIDOfTheVMContributorRoleAsSeenFromIAM->Roles"
  principal_id       = "1234-abcd-5678-efg"

schedule = {
start_date_time = time_static.example.rfc3339
expiration = {
duration_hours = 8
}
}
}
```
That should theoretically replicate the click-through process you described with just eligible-group.  I haven't touched PIM Groups yet but AFAIK that's not yet available in TF.  I'm told that the [AzAPI resource provider](https://registry.terraform.io/providers/Azure/azapi/latest/docs) will allow you to apply regular Azure API commands via TF, so if you know how to do that (I don't) it might be doable.

My lack of luck has been around the fact that we don't want to have expirations for most of our roles - the idea is that the role will be part of the solution that's in place, it never needs to expire since it's going to apply to everyone that's in the security group as they come and go.  Hopefully others have been successful.

**Comment by Quanalogy** [comment]
@MohnJadden 

I don't really get why you're talking about "resource groups" when what I'm looking for is purely in the AD.
When I do the before mentioned steps to make PIM for groups I do not use it in any resource group - I integrate with other systems that look into the AD and by membership within an AD group provide you a given access level. 

Also when you talk about "VM contributor", once again this really doesn't have anything to do with my usecase. Sorry if I overlook something here but it seems to me like being two completely different usecases.

**Comment by MohnJadden** [comment]
@Quanalogy  

The `azurerm_pim_eligible_role_assignment` TF resource provider covers PIM for Azure resources, not Azure AD roles or groups.  Since Azure resources are in resource groups, I used that as an example.  

I don't think Terraform yet supports PIM groups or PIM for AAD roles.  [There's a feature request for the AAD role PIM functionality](https://github.com/hashicorp/terraform-provider-azuread/issues/1164) if you wanted to add on the group functionality request.

**Comment by ZarakiKenpachi7** [comment]
@oWretch , @manicminer  - are there any open issues around configuring PIM for Azure Resources - **Role Settings** please? I believe it would be really nice to have role settings configuration via terraform as well with the already existing pim_assignment resources.

**Comment by MohnJadden** [comment]
@ZarakiKenpachi7 I opened one last week - https://github.com/hashicorp/terraform-provider-azurerm/issues/23458  

I had run into a very generic error when trying to assign a permanent role to a scope that had the default 1 year maximum, so having the role settings managed by TF will help greatly.  Feel free to throw a +1 and have some other folks do so as well; hopefully that helps give role settings some traction.

**Comment by davidng1996** [comment]
I have issues where PIM assignments are created but Terraform fails to detect the resources. This issue was raised in https://github.com/hashicorp/terraform-provider-azurerm/issues/23366 recently.

I have raised this few months back as well but it is not fixed - https://github.com/hashicorp/terraform-provider-azurerm/issues/22909

**Comment by Satak** [comment]
Just want to say that this Entra ID role PIM support for Terraform Azure AD provider is really needed and waited feature in our organization. I wish we could have some ETA? There are over 400 up votes for this issue so I'm assuming our organization is not the only one waiting for this. 🤞 🙏

**Comment by tim-chaffin** [comment]
PIM for Groups in my opinion, in general, is a challenging feature to automate, even without Terraform. From trying to get this to work, using PowerShell, or a chain of API calls from Postman, I've ended up with this workflow, which I'm sharing here as others may glean at least some automation on this front.

At the core of the automation, the following happens:

- Create two Entra ID Security Groups: 
  - The first group may already be present in your tenant, but its main purpose is to be a logical organization of team members who will be eligible to activate a PIM assignment, thus elevating their privileges.
  - The second group will be "Onboarded" as a PIM group later on.
  - Ensure that the `Microsoft Entra roles can be assigned to the group` flag is enabled on both groups. _Note: Your identity will need the `Privileged Role Administrator` role, as well as a P1 or P2 license in order to tick this flag._
- Configure the second group (our future PIM group) so that the first group (our team group) is a `Member` of the second group. [doc](https://learn.microsoft.com/en-us/entra/fundamentals/how-to-manage-groups#add-members-or-owners-of-a-group)
- Then use the [unifiedRoleManagementPolicy](https://learn.microsoft.com/en-us/graph/api/resources/unifiedrolemanagementpolicy?view=graph-rest-1.0) Graph API endpoint to `PATCH` the second group, with appropriate settings.
  - This is the weirdest thing about PIM for Groups. "You can't onboard a group to PIM for groups explicitly." _[source](https://learn.microsoft.com/en-us/graph/api/resources/privilegedidentitymanagement-for-groups-api-overview?view=graph-rest-1.0#onboarding-groups-to-pim-for-groups)_ But you can send a PIM for Groups setting call, which then in turn onboards the PIM group implicitly. **So confusing!**
  - Within the crazy lengthy rules payload, indicate that the first group has PIM settings on the second group per your discretion and standards.
  - Once this call is successfully completed, the second group turns into a PIM group with defined settings for the first group.
- Then use a standard Terraform `azurerm_role_assignment` resource block to associate the new PIM Group as the `principal_id` over any desired `scope` with privileges that match your standards.
- A user then logs in to Azure, and navigates to the `Microsoft Entra Privileged Identity Management` blade → My Roles → Groups → and activates membership into the newly configured PIM Group.

The challenge for Terraform right now, is that none of the [PIM for Groups API methods](https://learn.microsoft.com/en-us/graph/api/resources/privilegedidentitymanagement-for-groups-api-overview?view=graph-rest-1.0) are in the `azurerm` provider at this time. So, the only way to automate is by devising your own system to make these calls in an automated fashion.

**Comment by cpressland** [comment]
Looks like https://github.com/manicminer/hamilton/pull/277 got merged, any word on when this work is planned @oWretch?

**Comment by oWretch** [comment]
The PR for this is waiting review. @manicminer is away this week, so I'm hoping it will be released at the end of next week.

**Comment by kkadiyam** [comment]
> PIM for Groups in my opinion, in general, is a challenging feature to automate, even without Terraform. From trying to get this to work, using PowerShell, or a chain of API calls from Postman, I've ended up with this workflow, which I'm sharing here as others may glean at least some automation on this front.
> 
> At the core of the automation, the following happens:
> 
> * Create two Entra ID Security Groups:
>   
>   * The first group may already be present in your tenant, but its main purpose is to be a logical organization of team members who will be eligible to activate a PIM assignment, thus elevating their privileges.
>   * The second group will be "Onboarded" as a PIM group later on.
>   * Ensure that the `Microsoft Entra roles can be assigned to the group` flag is enabled on both groups. _Note: Your identity will need the `Privileged Role Administrator` role, as well as a P1 or P2 license in order to tick this flag._
> * Configure the second group (our future PIM group) so that the first group (our team group) is a `Member` of the second group. [doc](https://learn.microsoft.com/en-us/entra/fundamentals/how-to-manage-groups#add-members-or-owners-of-a-group)
> * Then use the [unifiedRoleManagementPolicy](https://learn.microsoft.com/en-us/graph/api/resources/unifiedrolemanagementpolicy?view=graph-rest-1.0) Graph API endpoint to `PATCH` the second group, with appropriate settings.
>   
>   * This is the weirdest thing about PIM for Groups. "You can't onboard a group to PIM for groups explicitly." _[source](https://learn.microsoft.com/en-us/graph/api/resources/privilegedidentitymanagement-for-groups-api-overview?view=graph-rest-1.0#onboarding-groups-to-pim-for-groups)_ But you can send a PIM for Groups setting call, which then in turn onboards the PIM group implicitly. **So confusing!**
>   * Within the crazy lengthy rules payload, indicate that the first group has PIM settings on the second group per your discretion and standards.
>   * Once this call is successfully completed, the second group turns into a PIM group with defined settings for the first group.
> * Then use a standard Terraform `azurerm_role_assignment` resource block to associate the new PIM Group as the `principal_id` over any desired `scope` with privileges that match your standards.
> * A user then logs in to Azure, and navigates to the `Microsoft Entra Privileged Identity Management` blade → My Roles → Groups → and activates membership into the newly configured PIM Group.
> 
> The challenge for Terraform right now, is that none of the [PIM for Groups API methods](https://learn.microsoft.com/en-us/graph/api/resources/privilegedidentitymanagement-for-groups-api-overview?view=graph-rest-1.0) are in the `azurerm` provider at this time. So, the only way to automate is by devising your own system to make these calls in an automated fashion.

Can someone please confirm , How to use "Then use the [unifiedRoleManagementPolicy](https://learn.microsoft.com/en-us/graph/api/resources/unifiedrolemanagementpolicy?view=graph-rest-1.0) Graph API endpoint to PATCH the second group, with appropriate settings" ?

I am in the same boat, I need to enable PIM for groups, After creating the AAD Group would like to run. powershell with unified policy to have mimil updates any policy changes and enable the PIM for groups

As provided in the documentation, I did updated to have scoped and scope type, But still doesn't work, running into provider missing errror

https://learn.microsoft.com/en-us/graph/api/resources/unifiedrolemanagementpolicy?view=graph-rest-1.0

Below is my code


$policies = Get-MgPolicyRoleManagementPolicyAssignment

# Placeholder for the right policy ID
$policyId = $null

foreach ($policy in $policies) {

}

if ($null -eq $policyId) {

} else {
    # Define minimal settings to update PIM policy
    $params = @{
        rules = @(
            @{
                "@odata.type" = "#microsoft.graph.unifiedRoleManagementPolicyApprovalRule"
                id = "Minimal_Changes_Approval"
                target = @{
                    caller = "EndUser"
                    operations = @("All")
                }
                level = "Assignment"
                setting = @{
                    isApprovalRequired = $false
                    isRequestorJustificationRequired = $true
                }
                # scopeId = "my-group-id"
                # scopeType= "Group"
            }
        )
    }

    # Update the unified role management policy with minimal changes
    Update-MgPolicyRoleManagementPolicy -UnifiedRoleManagementPolicyId $policyId -BodyParameter $params -scopeId = "04abf3cb-fec5-4bd5-87b3-7a7a08efcecd" -scopeType= "Group"

**Comment by MathewJohnsonIPO** [comment]
At this point you are probably better to wait until [this](https://github.com/hashicorp/terraform-provider-azuread/pull/1327) PR gets merged and released. I doubt it will be long now...

**Comment by tim-chaffin** [comment]
Its interesting they're moving it to AzureAD, where the other PIM stuff lives in AzureRM.

**Comment by MohnJadden** [comment]
@tim-chaffin I'm hoping the AzureAD part is meant to handle Azure AD/Entra ID roles, while AzureRM will handle Azure resource roles.

I'm still running into issues with the AzureRM counterpart but that's a whole other resource provider.

**Comment by oWretch** [comment]
@MohnJadden you are correct. hashicorp/terraform-provider-azuread#1327 is there to handle PIM assignments of one group to another Entra ID group. There is already support (with an outstanding issue) for Entra ID roles in the Azure AD provider. And there are probably fixes required for the Azure role assignments, as you have mentioned.

**Comment by hales8181** [comment]
Do we have an ETA on this at all?  I can't tell you how much I am eagerly awaiting this!

**Comment by MohnJadden** [comment]
I just @ed the one remaining reviewer in the other thread.  I'm not with HCP but hopefully this gets done and working before the acquisition completes and everyone cashes out.

**Comment by tim-chaffin** [comment]
![8p73sg](https://github.com/hashicorp/terraform-provider-azuread/assets/128724533/442a290c-c69a-43dc-87ab-8e651681bb9c)

**Comment by tim-chaffin** [comment]
@manicminer I saw that we've merged the much anticipated PR. This is excellent news! I know I for am very grateful for this.

However, can anyone shed some light on how these modules work? I checked the [docs](https://registry.terraform.io/providers/hashicorp/azuread/latest/docs), and I didn't see anything yet. If someone could help us see how we'd create a PIM group through these modules, and configure the attributes, that would be amazing.

**Comment by hales8181** [comment]
It's not been released yet. You need to wait for 2.49.0 to drop

https://github.com/hashicorp/terraform-provider-azuread/releases

**Comment by davidng1996** [comment]
Looks like this is targeting AD group only? Can it be used to target SP/User level? On 9 May 2024, at 12:33 AM, Tom Bamford ***@***.***> wrote:﻿
Yep that'll be released tomorrow. In the meantime, the docs are in the repo and have good examples to start from (if you wanted to build from source in the meantime).

**Comment by karts499** [comment]
Spread the news!

**Comment by jakubigla** [comment]
So here's the thing. Based on the documentation and the fact that I can do this via Portal without any of the `RoleManagementPolicy.ReadWrite.AzureADGroup` type of the permissions, an owner of the group can enable the PIM for groups without extra permissions. Anyone had a luck to use API based on the above conditions?
